presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -0,0 +1,159 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cassandra;
+
+import com.facebook.presto.cassandra.util.CassandraCqlUtils;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.predicate.Domain;
+import com.facebook.presto.spi.predicate.Range;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.cassandra.util.CassandraCqlUtils.toCQLCompatibleString;
+import static com.google.common.collect.Sets.cartesianProduct;
+import static java.util.Objects.requireNonNull;
+
+public class CassandraClusteringPredicatesExtractor
+{
+    private final List<CassandraColumnHandle> clusteringColumns;
+    private final TupleDomain<ColumnHandle> predicates;
+    private final ClusteringPushDownResult clusteringPushDownResult;
+
+    public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> clusteringColumns, TupleDomain<ColumnHandle> predicates)
+    {
+        this.clusteringColumns = ImmutableList.copyOf(requireNonNull(clusteringColumns, "clusteringColumns is null"));
+        this.predicates = requireNonNull(predicates, "predicates is null");
+        this.clusteringPushDownResult = getClusteringKeysSet(clusteringColumns, predicates);
+    }
+
+    public List<String> getClusteringKeyPredicates()
+    {
+        final Set<List<Object>> pushedDownDomainValues = clusteringPushDownResult.getDomainValues();
+
+        if (pushedDownDomainValues.isEmpty()) {
+            return ImmutableList.of();
+        }
+
+        ImmutableList.Builder<String> clusteringPredicates = ImmutableList.builder();
+        for (List<Object> clusteringKeys : pushedDownDomainValues) {
+            StringBuilder stringBuilder = new StringBuilder();
+
+            for (int i = 0; i < clusteringKeys.size(); i++) {
+                if (i > 0) {
+                    stringBuilder.append(" AND ");
+                }
+
+                stringBuilder.append(CassandraCqlUtils.validColumnName(clusteringColumns.get(i).getName()));
+                stringBuilder.append(" = ");
+                stringBuilder.append(CassandraCqlUtils.cqlValue(toCQLCompatibleString(clusteringKeys.get(i)), clusteringColumns.get(i).getCassandraType()));
+            }
+            if (!stringBuilder.toString().isEmpty() && !stringBuilder.toString().equals("")) {
```

Comment:
suggestion: instead of doing this ugly check do `continue` at the beginning of the loop if the clustering keys are empty.